packages/block-library/src/navigation-menu/block-navigator.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	Fragment,
+	useState,
+} from '@wordpress/element';
+import {
+	useSelect,
+	useDispatch,
+} from '@wordpress/data';
+import {
+	BlockNavigationList,
+} from '@wordpress/block-editor';
+import {
+	IconButton,
+	SVG,
+	Path,
+	Modal,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+const NavigatorIcon = (
+	<SVG xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width="20" height="20">
+		<Path d="M5 5H3v2h2V5zm3 8h11v-2H8v2zm9-8H6v2h11V5zM7 11H5v2h2v-2zm0 8h2v-2H7v2zm3-2v2h11v-2H10z" />
+	</SVG>
+);
+
+export default function BlockNavigator( { clientId } ) {
+	const [ isNavigationListOpen, setIsNavigationListOpen ] = useState( false );
+
+	const {
+		block,
+		selectedBlockClientId,
+	} = useSelect( ( select ) => {
+		const {
+			getSelectedBlockClientId,
+			getBlock,
+		} = select( 'core/block-editor' );
+
+		return {
+			block: getBlock( clientId ),
+			selectedBlockClientId: getSelectedBlockClientId(),
+		};
+	}, [ clientId ] );
+
+	const {
+		selectBlock,
+	} = useDispatch( 'core/block-editor' );
+
+	return (
+		<Fragment>
+			{ isNavigationListOpen && (
+				<Modal
+					title={ __( 'Block Navigator' ) }
+					closeLabel={ __( 'Close' ) }
+					onRequestClose={ () => setIsNavigationListOpen( false ) }
+				>
+					<BlockNavigationList
+						blocks={ [ block ] }
+						selectedBlockClientId={ selectedBlockClientId }
+						selectBlock={ selectBlock }
```

Comment:
Selecting the block works well in the current Block Navigation Menu accessible from the top toolbar. The PopOver closes and the correct block becomes selected.

In a modal this makes less sense. I think this is the biggest issue with the experience as it stands.

Reply:
Oh, I suppose it makes sense that `selectBlock` is causing the focus to shift and that's closing the modal.

Reply:
The navigator view is not meant for navigation only, it is, or it should be a way to build and edit the menu. We could make a double click action or some other way to select a deeply nested item and return to the navigation block edit view, for example a la dev tools "inspect" :)

Reply:
I'm not sold on using a modal for this functionality as it stands. What @draganescu is describing - that it should be a way to build and edit the menu - would make more sense as a modal, but on the other hand shouldn't the block itself provide that drag/drop/edit capability already?
If this navigator is replicating the top one, it would be better to have it as a dropdown, exactly like the top one. 

Another thing I notice with the modal is when closing it with `esc` focus goes to the block wrapper (not sure if that's what it's called - I mean the element that gets focus when navigating blocks with the arrow keys) whereas I would expect it to go back to the initial button, or at least to the next focusable element.

Reply:
The idea is definitely that this is a data view, and will be used for re-arranging things, potentially adding new blocks. I plan to do this incrementally though and not on the same PR. It's a good question about where navigation fits into this modal. 

> Another thing I notice with the modal is when closing it with esc focus goes to the block wrapper (not sure if that's what it's called - I mean the element that gets focus when navigating blocks with the arrow keys) whereas I would expect it to go back to the initial button, or at least to the next focusable element.

Good catch, will see if I can figure out what's going on.

Reply:
`esc` was indeed triggering navigation mode, and with the way the escape key event was working for modals there was no way of stopping propagation.

I've re-arranged things in https://github.com/WordPress/gutenberg/pull/17265/commits/fc2d2993895f8716ba46aed09cde6503b73c9332, and it's an improvement, but it still seems like the wrong icon is highlighted.

I'll have to bring those changes to the modal to a separate PR and will continue looking into the issue.

Reply:
Hmm yeah, seems like focus is always returned to the previous focusable element rather than the next one. Agree it's a lot better than it was 😄 
Apart from this modal the same problem was also affecting the one that pops up for "Resolve" on an invalid block. That one is also now returning focus to the previous focusable element.

Reply:
I have a fix locally and some general modal fixes on https://github.com/WordPress/gutenberg/pull/17297. Will bring them all together once that's merged.

Had a look at the that invalid block modal, and it looks like when the modal is open, the button that opens it is no longer rendered, so I think that causes the focus return not to work.